### PR TITLE
feat: implemented exclusive as whitelist 

### PR DESCRIPTION
**Pull request by mpark63**
## tl;dr
- exclusive is a string[] 
- each string is a whitelisted distribution that can overlap with given distribution 
- also applies to fine requirements 

## Motivation 
Until now, exclusive has been a boolean field of distribution and fine requirement objects.
- `exclusive` is true: exclusive to ALL, cannot double count with any 
- `exclusive` is false: exclusive to NONE, can double count with any 

But `exclusive` as a boolean cannot provide logic for 'gray zones' where a distribution is exclusive to some but not all distributions. Take any KSAS degree, for example. The `Humanities` distribution can overlap with the `Writing Intensive` distribution but _cannot_ overlap with the `Social Sciences`. Exclusive can be neither `true` nor `false`. 

## Implementation
- Edited type definitions for `requirement` of `distributionFunctions.js` and `FineReq` and `DistributionObj` of `commonTypes.tsx`
- Added exclusive fields on for all majors in `majors.tsx` e.g.
Screen Shot 2022-07-22 at 6 24 44 PM
- Modified the `updateReqs` function in `distributionBarsJSX.tsx` 

### The new exclusive field 
- when exclusive === undefined (usually because it's optional), the distribution can overlap with any others 
- when exclusive === ['N/A'], it may not overlap with any other distributions 
- when exclusive === ['A', 'B'], it may overlap with distributions with the `name` field === 'A' or 'B' 
- for fine requirements, we use the 'description' field instead of name 

## Testing 
Tested locally on the backend dev branch and frontend exclusive branch. Would like to test on our beta deployments.
Some interesting test cases were Computer Science and Cognitive Science.

### CODE REVIEW

File: lib/resources/majors.tsx [hunk 1]
```diff
@@ -2059,6 +2256,7 @@ const bsCS_Old: Major = {
             '<b>Intro Algorithms</b> <br /> EN.601.433 Intro Algorithms',
           required_credits: 3,
           criteria: 'EN.601.433[C]',
+          exclusive: ['N/A'],
```
Comment (aryamanshodhan) [inline lib/resources/majors.tsx:2259]: Do we need exclusive: ['N/A']? If there is no exclusivity, isn't it fine to just not specify the field because it is optional anyway?

**Review by aryamanshodhan — CHANGES_REQUESTED** [review]
One comment mentioned. Once you address it, you can assume that my PR is approved without explicitly asking me to review again.

I also assume that you whitelisted the correct distributions for each of the instances of exclusive in the majors; syntax looks fine for this.

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx [hunk 2]
```diff
@@ -192,39 +192,44 @@ const DistributionBarsJSX: FC<{ major: Major }> = ({ major }) => {
...
-        }
-      }),
-    );
+    // exclusive check:
+    // If exclusive is undefined, the course may double count for any number of distributions
+    // If exclusive is string[], the specified distributions / fine requirements are 'whitelisted'
+    // When a course satisfies a distribution, exclusive takes the value of distribution.exclusive
+    if (!courseObj) return;
+    let distExclusive: string[] | undefined = undefined; // initial value
+    reqs.forEach((reqGroup, i) => {
```
Comment (mattliu-mygit) [inline lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:203]: Trivial comment: this is O(N^2). Is there a better way to do this?

**Review by mattliu-mygit — APPROVED** [review]
Nice! Refer to my comment earlier.

**Review by kyang117 — APPROVED** [review]
thumbs up

File: lib/resources/majors.tsx (hunk 1 above)
Comment (mpark63) [reply lib/resources/majors.tsx:2259]: Ah I'm glad you brought it up. 
These are the possible values for `exclusive`: 
- undefined, always allows double count 
- ['N/A'], never allows double count 
- ['A', 'B'], it may overlap with distributions with the name field === 'A' or 'B'

In the case of `Algorithms`, 
- cannot overlap with `Upper Level CS Credits` or `CS Electives` => cannot be `undefined`
- can overlap with remaining 5 fine requirements => 
`exclusive: ['Gateway Computing: JAVA', 'Intermediate Programming', 'Data Structures', 'Computer System Fundamentals', 'Automata & Computation Theory', 'Team Requirement']`

Algo **never** satisfies these fine requirements. I give exclusive the value`['N/A']` so that no fine requirement is in the whitelist and we never bother running `checkRequirementSatisfied`. I do this only when I'm 100% certain a requirement never overlaps with another. 

Pros: 
- short circuiting logic, less function calls 
- saves us typing 

Cons: 
- inconsistent practice (sometimes I do this, other times I don't) 
- might introduce human error 

Hope that makes sense. I'm open to any suggestions. Lmk what you think. @aryamanshodhan @mattliu-mygit @kyang117

File: lib/components/dashboard/degree-info/DistributionBarsJSX.tsx (hunk 2 above)
Comment (mpark63) [reply lib/components/dashboard/degree-info/DistributionBarsJSX.tsx:203]: I don't think so, because reqs is a 2d array :? Is there a something I'm missing? 

Screen Shot 2022-07-23 at 9 02 23 PM
